### PR TITLE
ENH Adds no_batch_dim for NLLLoss

### PR DESCRIPTION
**Pull request by thomasjpfan**
Towards https://github.com/pytorch/pytorch/issues/60585

cc @albanD @mruberry @jbschlosser

### CODE REVIEW

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -164,10 +164,11 @@ class NLLLoss(_WeightedLoss):
             :attr:`reduction`. Default: ``'mean'``
 
     Shape:
-        - Input: :math:`(N, C)` where `C = number of classes`, or
+        - Input: :math:`(N, C)`, :math:`(C)`, where `C = number of classes`, or
```
Comment (jbschlosser) [inline torch/nn/modules/loss.py:167]: nit: order these shapes from least to most dimensions

File: torch/nn/modules/loss.py [hunk 2]
```diff
@@ -164,10 +164,11 @@ class NLLLoss(_WeightedLoss):
...
           :math:`(N, C, d_1, d_2, ..., d_K)` with :math:`K \geq 1`
           in the case of `K`-dimensional loss.
-        - Target: :math:`(N)` where each value is :math:`0 \leq \text{targets}[i] \leq C-1`, or
+        - Target: :math:`(N)` or a scalar, where each value is
```
Comment (jbschlosser) [inline torch/nn/modules/loss.py:170]: nit: Similarly here, order from least to most dimensions

File: aten/src/ATen/native/LossNLL.cpp [hunk 3]
```diff
@@ -22,10 +22,12 @@ TORCH_META_FUNC(nll_loss_forward)
   TORCH_CHECK(
       self.dim() > 0 && self.dim() <= 2, "input tensor should be 1D or 2D");
   TORCH_CHECK(
-      target.dim() == 1,
-      "1D target tensor expected, multi-target not supported");
+      target.dim() <= 1,
+      "0D or 1D target tensor expected, multi-target not supported");
```
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp:26]: nit: I think this should read "scalar or 1D target tensor..." instead

File: aten/src/ATen/native/LossNLL.cpp [hunk 4]
```diff
@@ -66,10 +68,12 @@ TORCH_META_FUNC(nll_loss_backward)
...
```
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp:72]: nit: similarly here, "scalar" instead of "0D"

**Review by jbschlosser — COMMENTED** [review]
I wonder if this should support both 0D targets and scalars?

**Comment by thomasjpfan** [comment]
> I wonder if this should support both 0D targets and scalars?

This PR adds support 0D targets and 1D inputs. Can you expand on your suggestion?

**Comment by jbschlosser** [comment]
> This PR adds support 0D targets and 1D inputs. Can you expand on your suggestion?

Sorry, should have been clearer. I mean should it support these cases:
* 1D input with a 0D target (e.g. `torch.tensor(2)`)
* 1D input with a scalar target (e.g. `2`)

Technically, 0D tensors and scalars are not equivalent- I'm leaning toward: it's fine to support only 0D tensors to avoid the complexity of dealing with additional scalar overloads.

**Comment by thomasjpfan** [comment]
As discussed during our meeting, this PR will only implement support for 0D tensors.

File: torch/nn/modules/loss.py [hunk 5]
```diff
@@ -164,10 +164,11 @@ class NLLLoss(_WeightedLoss):
             :attr:`reduction`. Default: ``'mean'``
 
     Shape:
-        - Input: :math:`(N, C)` where `C = number of classes`, or
+        - Input: :math:`(C)` or :math:`(N, C)`, where `C = number of classes`, or
           :math:`(N, C, d_1, d_2, ..., d_K)` with :math:`K \geq 1`
           in the case of `K`-dimensional loss.
-        - Target: :math:`(N)` where each value is :math:`0 \leq \text{targets}[i] \leq C-1`, or
+        - Target: A scalar or :math:`(N)`, where each value is
```
Comment (jbschlosser) [inline torch/nn/modules/loss.py:170]: nit: instead of "scalar", this should read shape `()` or `(N)`

**Review by jbschlosser — COMMENTED** [review]
Looks great! One minor doc comment to cement the "0D but not scalar" support

**Review by jbschlosser — APPROVED** [review]
LGTM!
